### PR TITLE
Example Gallery

### PR DESCRIPTION
**Pull request by rowanc1**
# Major changes
* updates to example documentation and build process
* no longer have to run the `__init__.py` when we add an example
* examples are now in their own folder with headers

## Minor changes
* Added `requirements_dev.txt`
* Added a `plotLayer` function in the `PlotUtils`
* improved the getting started pages

### Outstanding
- [ ] Merge the getting started/resources/tools from `tutorials`

The main issue this PR is based on is #554. This PR also addresses #553 #544 #491.

image

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -16,7 +16,6 @@ addons:
     - liblapack-dev
 
 env:
-    - TEST_DIR=tests/examples
     - TEST_DIR=tests/em/fdem/inverse/derivs
```
Comment (rowanc1) [inline .travis.yml]: I am no longer testing the examples. They are being tested inside of the doc tests, but will not go towards coverage.

File: .travis.yml [hunk 2]
```diff
@@ -46,11 +45,11 @@ before_install:
...
 install:
   - if [ ${TRAVIS_PYTHON_VERSION:0:1} == "2" ]; then
-      conda install --yes pip python=$TRAVIS_PYTHON_VERSION numpy scipy matplotlib cython ipython h5py nose sphinx sympy vtk;
+      conda install --yes pip python=$TRAVIS_PYTHON_VERSION numpy scipy matplotlib cython ipython h5py vtk;
     else
-      conda install --yes pip python=$TRAVIS_PYTHON_VERSION numpy scipy matplotlib cython ipython h5py nose sphinx sympy;
+      conda install --yes pip python=$TRAVIS_PYTHON_VERSION numpy scipy matplotlib cython ipython h5py;
     fi
-  - pip install nose-cov python-coveralls pymatsolver discretize
+  - pip install -r requirements_dev.txt
```
Comment (rowanc1) [inline .travis.yml]: I have put these inside the `req_dev`.

File: SimPEG/Utils/PlotUtils.py [hunk 3]
```diff
@@ -67,14 +68,61 @@ def plot2Ddata(xyz, data, vec=False, nx=100, ny=100,
...
         plt.axis("off")
         fig.savefig(figname, dpi=200)
-    if level is  None:
+    if level is None:
         return cont, ax
     else:
         return cont, ax, CS
+
+
+def plotLayer(sig, LocSigZ, xscale='log', ax=None,
```
Comment (rowanc1) [inline SimPEG/Utils/PlotUtils.py]: This was added from a @sgkang notebook.

File: docs/Makefile [hunk 4]
```diff
@@ -39,13 +39,20 @@ help:
 	@echo "  doctest    to run all doctests embedded in the documentation (if enabled)"
 
 clean:
-	-rm -rf $(BUILDDIR)/*
+	rm -rf $(BUILDDIR)/*
+	rm -rf content/examples/
```
Comment (rowanc1) [inline docs/Makefile]: These are added due to the gallery.

File: docs/conf.py [hunk 5]
```diff
@@ -28,12 +28,15 @@
...
-              'sphinx.ext.intersphinx',
-              'matplotlib.sphinxext.plot_directive']
+extensions = [
+    'sphinx.ext.todo',
+    'sphinx.ext.mathjax',
+    'sphinx.ext.viewcode',
+    'sphinx.ext.autodoc',
+    'sphinx.ext.intersphinx',
+    'matplotlib.sphinxext.plot_directive',
+    'sphinx_gallery.gen_gallery'
```
Comment (rowanc1) [inline docs/conf.py]: Look, a gallery!

File: docs/conf.py [hunk 6]
```diff
@@ -49,7 +52,7 @@
 
 # General information about the project.
 project = u'SimPEG'
-copyright = u'2013 - 2016, SimPEG Developers'
+copyright = u'2013 - 2017, SimPEG Developers'
```
Comment (rowanc1) [inline docs/conf.py]: because, it is now 2017!

File: docs/content/api_core/api_Maps.rst [hunk 7]
```diff
@@ -61,10 +61,10 @@ done by the :class:`SimPEG.Maps.ExpMap` described above.
     m = np.r_[0.2,1,0.1,2,2.9] # only 5 model parameters!
     sig = myMap * m
 
-.. plot::
 
-    from SimPEG import Examples
-    Examples.Maps_ComboMaps.run()
+.. image:: /content/examples/maps/images/sphx_glr_plot_combo_001.png
```
Comment (rowanc1) [inline docs/content/api_core/api_Maps.rst]: This is the way that I have gone around running the example twice in the doc build. Just have an image and link through to the example. The test should catch any broken links.

File: docs/content/api_core/api_installing.rst [hunk 8]
```diff
@@ -6,7 +6,7 @@ Getting Started with SimPEG
 Dependencies
 ============
 
-- Python 2.7
+- Python 3.5 (or 2.7)
```
Comment (rowanc1) [inline docs/content/api_core/api_installing.rst]: Minor updates. More can come in through the tutorials materials, however, some of those are now out of date.

File: examples/dc/plot_analytic_dipole.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
+"""
```
Comment (rowanc1) [inline examples/dc/plot_analytic_dipole.py]: I experimented with turning this one into a pure script. The rest I have left as `run()` in the `ifmain`.

File: examples/dc/plot_dc_schlumberger_array.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
+"""
```
Comment (rowanc1) [inline examples/dc/plot_dc_schlumberger_array.py]: This one I experimented with their "notebook" like format. It might be better to use a different plugin for these larger walk throughs...

File: examples/nsem/plot_foward_3D.py [hunk 11]
```diff
@@ -1,6 +1,11 @@
-# Test script to use SimPEG.MT platform to forward model synthetic data.
+"""
+MT: 3D: Forward
```
Comment (rowanc1) [inline examples/nsem/plot_foward_3D.py]: @grosenkj you will have to fight with a few conflicts no doubt. Hopefully it makes sense what I have moved around.

File: examples/utils/plot_surface2ind_topo.py [hunk 12]
```diff
@@ -1,3 +1,10 @@
+"""
+Utils: surface2ind_topo
```
Comment (rowanc1) [inline examples/utils/plot_surface2ind_topo.py]: Let me know if the layout/folder structure should be tweaked.

Comment (rowanc1) [inline examples/utils/plot_surface2ind_topo.py]: image

Comment (lheagy) [inline examples/utils/plot_surface2ind_topo.py]: If it is easy to change the order (and not have numbers in front of the names - that starts getting cluttered), I would suggest
- Basic Inversion
- Mesh Generation
- Mappings
- Gravity
- Magnetics
- Direct Current Resistivity
- Frequency Domain Electromagnetics
- Natural Source Electromagnetics (instead of Magnetotellurics)
- Time Domain Electromagnetics
- Fluid Flow
- Utilities

(so basics, applications in ~ order of complexity / grouped with like things, utilities)

Comment (lheagy) [inline examples/utils/plot_surface2ind_topo.py]: irrespective of it we change the order, Magnetotellurics should be call Natural Source Electromagnetics

File: SimPEG/Examples/Maps_ComboMaps.py [hunk 13]
```diff
@@ -1,66 +0,0 @@
-from SimPEG import Mesh, Maps
```
Comment (rowanc1) [inline SimPEG/Examples/Maps_ComboMaps.py]: No, It is still there. I guess that the rename didn't work in git..

File: examples/flow/plot_richards_forward_1D.py [hunk 14]
```diff
@@ -1,3 +1,43 @@
...
+
+Figure (c) shows the saturation field over the entire experiment. Here
+you can see that the timestep is not constant over time (5 seconds
+at the start of the simulation, 60 seconds at the end). You can also
+see the effect of the highly conductive layer in the model between
+20 and 25 cm depth. The water drains straight through the conductive
+unit and piles up on the other side - advancing the fluid front
+faster than the other layers.
+
+Rowan Cockett - 21/12/2016
```
Comment (rowanc1) [inline examples/flow/plot_richards_forward_1D.py]: That would be great. Would like to do that in a different PR.

**Comment by rowanc1** [comment]
I don't think that it is easy to put these in a specific order. At least not after reading the docs.

We can't even put numbers in front `01-Basic Inversion`?!

image

Not quite sure what is going on.

**Comment by rowanc1** [comment]
Better:

image

**Comment by rowanc1** [comment]
Better:

image

**Comment by rowanc1** [comment]
I think this is good once tests pass.

**Comment by lheagy** [comment]
nice work @rowanc1, this really bring the docs up a level :)
